/*
 *  This file is part of AndroidIDE.
 *
 *  AndroidIDE is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  AndroidIDE is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *   along with AndroidIDE.  If not, see <https://www.gnu.org/licenses/>.
 */

package com.itsaky.androidide.fragments.output;

import android.os.Bundle;
import android.view.View;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import com.itsaky.androidide.R;
import com.itsaky.androidide.databinding.FragmentNonEditableEditorBinding;
import com.itsaky.androidide.editor.ui.IDEEditor;
import com.itsaky.androidide.eventbus.events.preferences.PreferenceChangeEvent;
import com.itsaky.androidide.fragments.EmptyStateFragment;
import com.itsaky.androidide.preferences.internal.EditorPreferences;
import com.itsaky.androidide.syntax.colorschemes.SchemeAndroidIDE;
import com.itsaky.androidide.utils.BasicBuildInfo;
import com.itsaky.androidide.utils.TypefaceUtilsKt;
import io.github.rosemoe.sora.lang.EmptyLanguage;
import org.greenrobot.eventbus.EventBus;
import org.greenrobot.eventbus.Subscribe;
import org.greenrobot.eventbus.ThreadMode;

public abstract class NonEditableEditorFragment extends
		EmptyStateFragment<FragmentNonEditableEditorBinding>
		implements ShareableOutputFragment, WrappableOutputFragment {

	public NonEditableEditorFragment() {
		super(R.layout.fragment_non_editable_editor, FragmentNonEditableEditorBinding::bind);
	}

	@Override
	public void clearOutput() {
		final var editor = getEditor();
		if (editor == null) {
			return;
		}

		// Editing CodeEditor's content is a synchronized operation
		editor.getText().delete(0, editor.getText().length());
		getEmptyStateViewModel().setEmpty(true);
	}

	@Nullable
	public IDEEditor getEditor() {
		final var binding = get_binding();
		if (binding == null) {
			return null;
		}
		return binding.editor;
	}

	@NonNull
	@Override
	public String getShareableContent() {
		final var editor = getEditor();
		if (editor == null) {
			return "";
		}

		final var editorText = editor.getText().toString();
		return BasicBuildInfo.shareableBuildInfo() + System.lineSeparator() + editorText;
	}

	@NonNull
	@Override
	public String getShareableFilename() {
		return "build_output";
	}

	@Override
	public boolean isWordWrapEnabled() {
		IDEEditor editor = getEditor();
		return editor != null && editor.isWordwrap();
	}

	@Override
	public void onDestroyView() {
		if (EventBus.getDefault().isRegistered(this)) {
			EventBus.getDefault().unregister(this);
		}
		super.onDestroyView();
	}

	@Subscribe(threadMode = ThreadMode.MAIN)
	public void onPreferenceChanged(PreferenceChangeEvent event) {
		if (EditorPreferences.OUTPUT_WORD_WRAP.equals(event.getKey())) {
			setWordWrapEnabled(Boolean.TRUE.equals(event.getValue()));
		}
	}

	@Override
	public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
		super.onViewCreated(view, savedInstanceState);
		if (!EventBus.getDefault().isRegistered(this)) {
			EventBus.getDefault().register(this);
		}
		getEmptyStateViewModel().setEmptyMessage(createEmptyStateMessage());
		final var editor = getBinding().editor;
		editor.setEditable(false);
		editor.setDividerWidth(0);
		editor.setEditorLanguage(new EmptyLanguage());
		editor.setWordwrap(EditorPreferences.INSTANCE.getOutputWordWrap());
		editor.setUndoEnabled(false);
		editor.setTypefaceLineNumber(TypefaceUtilsKt.jetbrainsMono());
		editor.setTypefaceText(TypefaceUtilsKt.jetbrainsMono());
		editor.setTextSize(12);
		editor.setColorScheme(SchemeAndroidIDE.newInstance(requireContext()));
	}

	@Override
	public void setWordWrapEnabled(boolean enabled) {
		IDEEditor editor = getEditor();
		if (editor != null) {
			editor.setWordwrap(enabled);
		}
	}

	@NonNull
	private CharSequence createEmptyStateMessage() {
		return "";
	}
}
